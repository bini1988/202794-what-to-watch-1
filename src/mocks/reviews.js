
export default [
  {
    id: 1,
    author: `Kate Muir`,
    rating: 8.9,
    text: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
    date: `2019-05-08T14:13:56.569Z`
  }, {
    id: 2,
    author: `Mollie`,
    rating: 3,
    text: `It was well acted, directed, and the music was good. But the story is yawn. Not trying to rip anybody but I checked my watch a dozen times during this movie.`,
    datetime: `2019-06-14T06:39:07.535Z`,
    date: `2019-06-14T06:39:07.535Z`
  }, {
    id: 3,
    author: `Mollie`,
    rating: 4.5,
    text: `I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.`,
    datetime: `2019-05-27T06:39:07.535Z`,
    date: `2019-05-27T06:39:07.535Z`
  }, {
    id: 4,
    author: `Mollie`,
    rating: 2,
    text: `I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.`,
    datetime: `2019-05-27T06:39:07.535Z`,
    date: `2019-05-27T06:39:07.535Z`
  }, {
    id: 5,
    author: `Mollie`,
    rating: 3,
    text: `It was well acted, directed, and the music was good. But the story is yawn. Not trying to rip anybody but I checked my watch a dozen times during this movie.`,
    datetime: `2019-06-14T06:39:07.535Z`,
    date: `2019-06-14T06:39:07.535Z`
  }, {
    id: 6,
    author: `Mollie`,
    rating: 5.0,
    text: `I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.`,
    datetime: `2019-05-27T06:39:07.535Z`,
    date: `2019-05-27T06:39:07.535Z`
  }, {
    id: 7,
    author: `Mollie`,
    rating: 2.5,
    text: `I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.`,
    datetime: `2019-05-27T06:39:07.535Z`,
    date: `2019-05-27T06:39:07.535Z`
  }
];
