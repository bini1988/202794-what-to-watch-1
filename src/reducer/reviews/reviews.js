import {NotificationManager} from "react-notifications";
import {normalizeItems} from "../../utils";

export const ActionTypes = {
  STORE_REVIEWS: `STORE_REVIEWS`,
};

export const ActionCreator = {
  /**
   * Сохранить список отзывов к фильму
   * @param {number} id ID фильма
   * @param {Object} items Map-объект отзывов
   * @param {string[]} itemsIds Список id's отзывов
   * @return {Object}
   */
  storeReviews: (id, items, itemsIds) => {
    return {type: ActionTypes.STORE_REVIEWS, meta: {id, itemsIds}, payload: items};
  },
};

export const Operation = {
  /**
   * Получить список отзывов к фильму
   * @param {number} id ID фильма
   * @return {Promise}
   */
  fetchMovieReviews: (id) => {
    return (dispath, getState, api) => {
      return api.fetchMovieReviews(id)
        .then((reviews) => {
          const {items, itemsIds} = normalizeItems(reviews);
          dispath(ActionCreator.storeReviews(id, items, itemsIds));
        }).catch((error) => {
          NotificationManager.error(error.message);
        });
    };
  },
  /**
   * Опубликовать отзыв к фильму
   * @param {number} id ID фильма
   * @param {Object} review Отзыв к фильму
   * @param {number} review.rating Оценка фильма
   * @param {string} review.comment Текст отзыва
   * @return {Promise}
   */
  submitMovieReview: (id, review) => {
    return (dispath, getState, api) => {
      return api.submitMovieReview(id, review)
        .then((reviews) => {
          const {items, itemsIds} = normalizeItems(reviews);
          dispath(ActionCreator.storeReviews(id, items, itemsIds));
        }).catch((error) => {
          NotificationManager.error(error.message);
          throw error;
        });
    };
  },
};


export const initialState = {
  /** Map-объект фильм -> массив отзывов */
  movieToItems: {},
  /** Map-объект отзывов */
  items: {},
};

export default (state = initialState, action) => {
  switch (action.type) {
    case ActionTypes.STORE_REVIEWS:
      return {
        ...state,
        movieToItems: {
          ...state.movieToItems,
          [action.meta.id]: action.meta.itemsIds,
        },
        items: {
          ...state.items,
          ...action.payload,
        },
      };
    default:
      return state;
  }
};
