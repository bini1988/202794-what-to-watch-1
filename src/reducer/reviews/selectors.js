import NameSpace from "../name-spaces";

/**
 * Получение таблицы отзывов к фильмам
 * @param {Object} state Текущее состояние redux стора
 * @return {Object[]}
 */
export const getReviews = (state) => {
  return state[NameSpace.Reviews].items;
};
