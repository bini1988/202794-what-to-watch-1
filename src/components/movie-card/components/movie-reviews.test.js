import React from "react";
import renderer from "react-test-renderer";
import MovieReviews from "./movie-reviews";
import card from "../../../mocks/movie-card";
import reviews from "../../../mocks/reviews";

it(`MovieReviews correctly renders default markup`, () => {
  const tree = renderer.create(
      <MovieReviews card={{...card, reviews}}/>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
