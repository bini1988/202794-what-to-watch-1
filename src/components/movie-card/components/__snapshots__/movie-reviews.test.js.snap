// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieReviews correctly renders default markup 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
      id="review-1"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Kate Muir
          </cite>
          <time
            className="review__date"
            dateTime="2019-05-08T14:13:56.569Z"
          >
            Wed May 08 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        1.5
      </div>
    </div>
    <div
      className="review"
      id="review-2"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          It was well acted, directed, and the music was good. But the story is yawn. Not trying to rip anybody but I checked my watch a dozen times during this movie.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Mollie
          </cite>
          <time
            className="review__date"
            dateTime="2019-06-14T06:39:07.535Z"
          >
            Fri Jun 14 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        2.0
      </div>
    </div>
    <div
      className="review"
      id="review-3"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Mollie
          </cite>
          <time
            className="review__date"
            dateTime="2019-05-27T06:39:07.535Z"
          >
            Mon May 27 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        2.5
      </div>
    </div>
    <div
      className="review"
      id="review-4"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Mollie
          </cite>
          <time
            className="review__date"
            dateTime="2019-05-27T06:39:07.535Z"
          >
            Mon May 27 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        3.0
      </div>
    </div>
  </div>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
      id="review-5"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          It was well acted, directed, and the music was good. But the story is yawn. Not trying to rip anybody but I checked my watch a dozen times during this movie.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Mollie
          </cite>
          <time
            className="review__date"
            dateTime="2019-06-14T06:39:07.535Z"
          >
            Fri Jun 14 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        3.5
      </div>
    </div>
    <div
      className="review"
      id="review-6"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Mollie
          </cite>
          <time
            className="review__date"
            dateTime="2019-05-27T06:39:07.535Z"
          >
            Mon May 27 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        4.0
      </div>
    </div>
    <div
      className="review"
      id="review-7"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          I love this movie. This film is a milestone in cinematography. Great Immersive camera-work. This film is an experience and i has already seen it 4 times and I only see more quality of the film.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Mollie
          </cite>
          <time
            className="review__date"
            dateTime="2019-05-27T06:39:07.535Z"
          >
            Mon May 27 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        5.0
      </div>
    </div>
  </div>
</div>
`;
